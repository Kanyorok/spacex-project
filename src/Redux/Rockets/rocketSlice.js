import { createSlice, createAsyncThunk } from '@reduxjs/toolkit';
import axios from 'axios';

const baseURL = 'https://api.spacexdata.com/v3/rockets';
export const displayRockets = createAsyncThunk('rockets', async () => {
  try {
    const retrievedRockets = await axios.get(baseURL);
    const rocketsDisplay = retrievedRockets.data.map((roc) => ({
      id: roc.id,
      name: roc.rocket_name,
      type: roc.rocket_type,
      description: roc.description,
      images: roc.flickr_images,
      reserved: false,
    }));
    return rocketsDisplay;
  } catch (error) {
    throw Error(error);
  }
});
const initialState = {
  rockets: [],
  loading: 'idle',
};
const rocketSlice = createSlice({
  name: 'rockets',
  initialState,
  reducers: {
    bookRocket: (state, action) => {
      const rocketId = action.payload;
      state.rockets = state.rockets.map((rocket) => (rocket.id === rocketId
        ? { ...rocket, reserved: true }
        : rocket));
    },
    cancelRocket: (state, action) => {
      const rocketId = action.payload;
      state.rockets = state.rockets.map((rocket) => (rocket.id === rocketId
        ? { ...rocket, reserved: false }
        : rocket));
    },
  },
  extraReducers: (builder) => {
    builder
      .addCase(displayRockets.pending, (state) => {
        state.loading = 'pending';
      })
      .addCase(displayRockets.fulfilled, (state, action) => {
        state.rockets = action.payload;
        state.loading = 'Succeeded loading rockets';
      })
      .addCase(displayRockets.rejected, (state) => {
        state.loading = 'failed load missions';
      });
  },
});
export const { bookRocket, cancelRocket } = rocketSlice.actions;
export default rocketSlice.reducer;
