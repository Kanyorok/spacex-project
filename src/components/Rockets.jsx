import React, { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { displayRockets } from '../Redux/Rockets/rocketSlice';

const Rockets = () => {
  const { rockets } = useSelector((state) => state.rockets);
  const dispatch = useDispatch();
  useEffect(() => {
    dispatch(displayRockets());
  }, [dispatch]);
  return (
    <div>
      {rockets.map((roc) => (
        <h1 key={roc.id}>
          id:
          {roc.id}
          name:
          {roc.name}
          type:
          {roc.type}
          images:
          {roc.flickr_images}
        </h1>

      ))}
    </div>
  );
};
export default Rockets;
