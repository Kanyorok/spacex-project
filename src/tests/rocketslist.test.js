import '@testing-library/jest-dom';
import { BrowserRouter } from 'react-router-dom';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import Rockets from '../components/Rockets';
import { displayRockets } from '../Redux/Rockets/rocketSlice';

jest.mock('../Redux/Rockets/rocketSlice', () => ({
  displayRockets: jest.fn(),
}));

const mockStore = configureStore([]);
const initialState = {
  rockets: {
    loading: false,
    rockets: [
      {
        id: '1',
        name: 'Falcon 1',
        description: 'It is the best film',
        reserved: true,
      },
      {
        id: '2',
        name: 'Mock Mission 2',
        description: 'It is the best film',
        reserved: false,
      },
    ],
  },
};
const store = mockStore(initialState);

test('Rockets component renders correctly', () => {
  displayRockets.mockReturnValueOnce({ type: 'mockAction' });
  const tree = renderer
    .create(
      <Provider store={store}>
        <BrowserRouter>
          <Rockets />
        </BrowserRouter>
      </Provider>,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
