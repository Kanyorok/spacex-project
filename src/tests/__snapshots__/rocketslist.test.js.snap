// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Missions component renders correctly 1`] = `
<div
  className="rock"
>
  <div
    className="test"
  >
    <img
      alt=""
      className="img"
    />
    <div
      className="info"
    >
      <h1>
        Falcon 1
      </h1>
      <span
        className="btn"
      >
        Reserved
      </span>
      <p>
        It is the best film
      </p>
      <button
        className="btn"
        onClick={[Function]}
        type="button"
      >
        Cancel Reservation
      </button>
    </div>
  </div>
  <div
    className="test"
  >
    <img
      alt=""
      className="img"
    />
    <div
      className="info"
    >
      <h1>
        Mock Mission 2
      </h1>
      <span
        className="btn"
      />
      <p>
        It is the best film
      </p>
      <button
        className="btn"
        onClick={[Function]}
        type="button"
      >
        Book Rocket
      </button>
    </div>
  </div>
</div>
`;

exports[`Rockets component renders correctly 1`] = `
<div
  className="rock"
>
  <div
    className="test"
  >
    <img
      alt=""
      className="img"
    />
    <div
      className="info"
    >
      <h1>
        Falcon 1
      </h1>
      <span
        className="btn"
      >
        Reserved
      </span>
      <p>
        It is the best film
      </p>
      <button
        className="btn"
        onClick={[Function]}
        type="button"
      >
        Cancel Reservation
      </button>
    </div>
  </div>
  <div
    className="test"
  >
    <img
      alt=""
      className="img"
    />
    <div
      className="info"
    >
      <h1>
        Mock Mission 2
      </h1>
      <span
        className="btn"
      />
      <p>
        It is the best film
      </p>
      <button
        className="btn"
        onClick={[Function]}
        type="button"
      >
        Book Rocket
      </button>
    </div>
  </div>
</div>
`;
